File: js/src/2025/Day1A.ts
import { readInputLinesFromPath } from '../utils/fileReader';
import * as path from 'path';

enum Direction {
    Left = 'L',
    Right = 'R'
}

interface Action {
    direction: Direction;
    clicks: number;
}

export class Day1A {
    private run(filePath: string): string {
        const actions = this.parse(filePath);
        
        let start = 50;
        let count = 0;
        
        for (const action of actions) {
            const clicks = action.clicks % 100;
            switch (action.direction) {
                case Direction.Left:
                    start = start - clicks;
                    break;
                case Direction.Right:
                    start = start + clicks;
                    break;
            }
            
            if (start < 0) {
                start = 100 - Math.abs(start);
            } else if (start > 99) {
                start = 0 + start - 100;
            }
            
            if (start === 0) {
                count += 1;
            }
        }
        
        return String(count);
    }
    
    private parse(filePath: string): Action[] {
        const lines = readInputLinesFromPath(filePath);
        return lines.map(line => {
            const dir = line[0] === 'L' ? Direction.Left : Direction.Right;
            const num = parseInt(line.slice(1), 10);
            return { direction: dir, clicks: num };
        });
    }
    
    execute(test: boolean): string {
        const testPath = path.join(__dirname, '../../tests/2025/');
        if (test) {
            return this.run(path.join(testPath, '1_test.txt'));
        } else {
            return this.run(path.join(testPath, '1.txt'));
        }
    }
}

File: aoc2025/Day1A.swift
import Foundation

fileprivate enum Direction {
    case left, right
}

class Day1A {
    private func run(path: String) -> String {
        let actions = parse(path: path)
        
        var start = 50
        var count = 0
        
        for action in actions {
            let clicks = action.clicks % 100
            switch action.direction {
            case .left:
                start = start - clicks
            case .right:
                start = start + clicks
            }
            
            if start < 0 {
                start = 100 - abs(start)
            } else if start > 99 {
                start = 0 + start - 100
            }
            
            if start == 0 {
                count += 1
            }
            print(start)
        }
        
        return String(count)
    }
    
    private func parse(path: String) -> [(direction: Direction, clicks: Int)] {
        let fileURL = URL(fileURLWithPath: path)
        let res = try! String(contentsOf: fileURL, encoding: .utf8).split(separator: "\n")
            .map {
                let dir = $0.first == "L" ? Direction.left : Direction.right
                let num = Int($0.dropFirst())!
                return (direction: dir, clicks: num)
            }

        return res
    }
}

extension Day1A: AoCTest {
    func execute(test: Bool) -> String {
        if test {
            return run(path: testPath + "1_test.txt")
        } else {
            return run(path: testPath + "1.txt")
        }
    }
}
